In [ ]:
import numpy as np

# Example 1

__a)__

$$
y = w \cdot x
$$

$$
MSE = \frac{1}{n} \sum_{i=1}^n (y_i - w \cdot x_i)^2 \to \min_{w}
$$

$$
\frac{\partial MSE}{\partial w} = -2 \frac{1}{n}  \sum_{i=1}^n (y_i - w \cdot x_i) \cdot x_i = 0
$$

Нормальное уравнение
$$
w= \frac{\sum_{i=1}^n y_i \cdot x_i}{\sum_{i=1}^n x_i^2}
$$

__b)__

$$
w_0=0,
x = (1, 2),
y  = (2, 3),
\eta=0.1,
$$

$$
w_1 =  w_0 - \eta \cdot \frac{\partial MSE}{\partial w}(w_0)\\
w_1 = 0 - 0.1 \cdot \frac{-2}{2} \cdot ((2 - 0 \cdot 1)\cdot 1 + (3 - 0 \cdot 2) \cdot 2) = 0.8 \\
w_2 = 0.8 - 0.1 \cdot \frac{-2}{2} \cdot ((2 - 0.8 \cdot 1)\cdot 1 + (3 - 0.8 \cdot 2) \cdot 2)=1.2
$$


In [ ]:
x = np.array([1.,2.])
y = np.array([2.,3.])
w=0.
n = 0.1
for i in range(3):
  w = w - n*(-2./2.)*((y[0]-w*x[0])*x[0] + ((y[1]-w*x[1])*x[1]))
  print(f'w_{i+1} = {w}')

w_1 = 0.8
w_2 = 1.2000000000000002
w_3 = 1.4000000000000001


$$
w_3 =  1.4
$$

__v)__

$$
w_1 = w_0 - 0.1 \cdot -2(2 - w_0 \cdot 1)\cdot 1\\
w_2 = w_1 - 0.1 \cdot -2(3 - w_1 \cdot 2)\cdot 2\\
w_3 = w_2 - 0.1 \cdot -2(2 - w_2 \cdot 1)\cdot 1\\
w_4 = w_3 - 0.1 \cdot -2(3 - w_3 \cdot 2)\cdot 2\\
$$

__g)__
